corehq/plugins/interface.py

```diff
@@ -0,0 +1,92 @@
+import inspect
+from collections import defaultdict
+
+import attr
+
+from dimagi.utils.logging import notify_exception
+from dimagi.utils.modules import to_function
+
+
+@attr.s(frozen=True)
+class ExtensionPoint:
+    name = attr.ib()
+    providing_args = attr.ib()
+    docs = attr.ib(default="")
+
+
+class PluginContribution:
+    def __init__(self, callable_ref, domains=None):
+        self.callable_ref = callable_ref
+        self.domains = domains
+        self._callable = None
+
+    def validate(self, extension_point):
+        _callable = self.callable_ref
+        if isinstance(_callable, str):
+            _callable = to_function(self.callable_ref)
+        if not _callable:
+            raise Exception(f"Plugin not found: '{self.callable_ref}'")
+        if not callable(_callable):
+            raise Exception(f"Plugin not callable: '{self.callable_ref}'")
+        self._callable = _callable
+        spec = inspect.getfullargspec(_callable)
+        unconsumed_args = set(extension_point.providing_args) - set(spec.args)
+        if unconsumed_args and not spec.varkw:
+            raise Exception(f"Not all extension point args are consumed: {unconsumed_args}")
+
+    def should_call(self, **kwargs):
+        if self.domains is None or 'domain' not in kwargs:
+            return True
+
+        return kwargs['domain'] in self.domains
+
+    def __call__(self, **kwargs):
+        if self.should_call(**kwargs):
+            return self._callable(**kwargs)
+
+    def __repr__(self):
+        return f"{self.callable_ref}"
+
+
+class Plugins:
```

Comment:
Non-blocking: would it make sense to call these things "Extensions" instead of "Plugins" since they consist of "extension points"? Similar to how Firefox and Chrome have extensions (Java and Flash were plugins, but they are being phased out).